// better-auth/dist/api.cjs
'use strict';

var chunkFTP5NPIZ_cjs = require('./chunk-FTP5NPIZ.cjs');
require('./chunk-NU4SXYJ5.cjs');
require('./chunk-V6RGXSG4.cjs');
require('./chunk-S5UORXJH.cjs');
require('./chunk-JOH6QBAP.cjs');
require('./chunk-UQC2FUXC.cjs');
require('./chunk-H74YRRNV.cjs');
require('./chunk-5E75URIA.cjs');
require('./chunk-CCKQSGIR.cjs');
require('./chunk-G2LZ73E2.cjs');
require('./chunk-OA7Z7IZT.cjs');
require('./chunk-2HPSCSV7.cjs');
require('./chunk-VXYIYABQ.cjs');
require('./chunk-PEZRSDZS.cjs');



Object.defineProperty(exports, "APIError", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.APIError; }
});
Object.defineProperty(exports, "callbackOAuth", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.callbackOAuth; }
});
Object.defineProperty(exports, "changeEmail", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.changeEmail; }
});
Object.defineProperty(exports, "changePassword", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.changePassword; }
});
Object.defineProperty(exports, "createAuthEndpoint", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.createAuthEndpoint; }
});
Object.defineProperty(exports, "createAuthMiddleware", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.createAuthMiddleware; }
});
Object.defineProperty(exports, "createEmailVerificationToken", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.createEmailVerificationToken; }
});
Object.defineProperty(exports, "deleteUser", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.deleteUser; }
});
Object.defineProperty(exports, "deleteUserCallback", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.deleteUserCallback; }
});
Object.defineProperty(exports, "error", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.error; }
});
Object.defineProperty(exports, "forgetPassword", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.forgetPassword; }
});
Object.defineProperty(exports, "forgetPasswordCallback", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.forgetPasswordCallback; }
});
Object.defineProperty(exports, "freshSessionMiddleware", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.freshSessionMiddleware; }
});
Object.defineProperty(exports, "getEndpoints", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.getEndpoints; }
});
Object.defineProperty(exports, "getSession", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.getSession; }
});
Object.defineProperty(exports, "getSessionFromCtx", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.getSessionFromCtx; }
});
Object.defineProperty(exports, "linkSocialAccount", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.linkSocialAccount; }
});
Object.defineProperty(exports, "listSessions", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.listSessions; }
});
Object.defineProperty(exports, "listUserAccounts", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.listUserAccounts; }
});
Object.defineProperty(exports, "ok", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.ok; }
});
Object.defineProperty(exports, "optionsMiddleware", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.optionsMiddleware; }
});
Object.defineProperty(exports, "originCheck", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.originCheck; }
});
Object.defineProperty(exports, "originCheckMiddleware", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.originCheckMiddleware; }
});
Object.defineProperty(exports, "resetPassword", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.resetPassword; }
});
Object.defineProperty(exports, "revokeOtherSessions", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.revokeOtherSessions; }
});
Object.defineProperty(exports, "revokeSession", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.revokeSession; }
});
Object.defineProperty(exports, "revokeSessions", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.revokeSessions; }
});
Object.defineProperty(exports, "router", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.router; }
});
Object.defineProperty(exports, "sendVerificationEmail", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.sendVerificationEmail; }
});
Object.defineProperty(exports, "sendVerificationEmailFn", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.sendVerificationEmailFn; }
});
Object.defineProperty(exports, "sessionMiddleware", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.sessionMiddleware; }
});
Object.defineProperty(exports, "setPassword", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.setPassword; }
});
Object.defineProperty(exports, "signInEmail", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.signInEmail; }
});
Object.defineProperty(exports, "signInSocial", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.signInSocial; }
});
Object.defineProperty(exports, "signOut", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.signOut; }
});
Object.defineProperty(exports, "signUpEmail", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.signUpEmail; }
});
Object.defineProperty(exports, "unlinkAccount", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.unlinkAccount; }
});
Object.defineProperty(exports, "updateUser", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.updateUser; }
});
Object.defineProperty(exports, "verifyEmail", {
  enumerable: true,
  get: function () { return chunkFTP5NPIZ_cjs.verifyEmail; }
});
