// better-auth/dist/api.js
export { APIError, callbackOAuth, changeEmail, changePassword, createAuthEndpoint, createAuthMiddleware, createEmailVerificationToken, deleteUser, deleteUserCallback, error, forgetPassword, forgetPasswordCallback, freshSessionMiddleware, getEndpoints, getSession, getSessionFromCtx, linkSocialAccount, listSessions, listUserAccounts, ok, optionsMiddleware, originCheck, originCheckMiddleware, resetPassword, revokeOtherSessions, revokeSession, revokeSessions, router, sendVerificationEmail, sendVerificationEmailFn, sessionMiddleware, setPassword, signInEmail, signInSocial, signOut, signUpEmail, unlinkAccount, updateUser, verifyEmail } from './chunk-HMECAGV7.js';
import './chunk-VXDYNCOX.js';
import './chunk-GTQM7JU7.js';
import './chunk-XFCIANZX.js';
import './chunk-WBFH3ZYN.js';
import './chunk-WZ6OIVJK.js';
import './chunk-KLDFBLYL.js';
import './chunk-NPO64SVN.js';
import './chunk-3XTQSPPA.js';
import './chunk-DBPOZRMS.js';
import './chunk-53IPO2PJ.js';
import './chunk-FURNA6HY.js';
import './chunk-TQQSPPNA.js';
import './chunk-UNWCXKMP.js';
