// better-auth/dist/chunk-IYXPTEHV.cjs
'use strict';

var chunkFTP5NPIZ_cjs = require('./chunk-FTP5NPIZ.cjs');
var chunkS5UORXJH_cjs = require('./chunk-S5UORXJH.cjs');
var chunkJOH6QBAP_cjs = require('./chunk-JOH6QBAP.cjs');
var chunkUQC2FUXC_cjs = require('./chunk-UQC2FUXC.cjs');

// src/plugins/anonymous/index.ts
var schema = {
  user: {
    fields: {
      isAnonymous: {
        type: "boolean",
        required: false
      }
    }
  }
};
var anonymous = (options) => {
  const ERROR_CODES = {
    FAILED_TO_CREATE_USER: "Failed to create user",
    COULD_NOT_CREATE_SESSION: "Could not create session",
    ANONYMOUS_USERS_CANNOT_SIGN_IN_AGAIN_ANONYMOUSLY: "Anonymous users cannot sign in again anonymously"
  };
  return {
    id: "anonymous",
    endpoints: {
      signInAnonymous: chunkFTP5NPIZ_cjs.createAuthEndpoint(
        "/sign-in/anonymous",
        {
          method: "POST",
          metadata: {
            openapi: {
              description: "Sign in anonymously",
              responses: {
                200: {
                  description: "Sign in anonymously",
                  content: {
                    "application/json": {
                      schema: {
                        type: "object",
                        properties: {
                          user: {
                            $ref: "#/components/schemas/User"
                          },
                          session: {
                            $ref: "#/components/schemas/Session"
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        },
        async (ctx) => {
          const { emailDomainName = chunkS5UORXJH_cjs.getOrigin(ctx.context.baseURL) } = options || {};
          const id = ctx.context.generateId({ model: "user" });
          const email = `temp-${id}@${emailDomainName}`;
          const newUser = await ctx.context.internalAdapter.createUser({
            id,
            email,
            emailVerified: false,
            isAnonymous: true,
            name: "Anonymous",
            createdAt: /* @__PURE__ */ new Date(),
            updatedAt: /* @__PURE__ */ new Date()
          });
          if (!newUser) {
            return ctx.json(null, {
              status: 500,
              body: {
                message: ERROR_CODES.FAILED_TO_CREATE_USER,
                status: 500
              }
            });
          }
          const session = await ctx.context.internalAdapter.createSession(
            newUser.id,
            ctx.request
          );
          if (!session) {
            return ctx.json(null, {
              status: 400,
              body: {
                message: ERROR_CODES.COULD_NOT_CREATE_SESSION
              }
            });
          }
          await chunkJOH6QBAP_cjs.setSessionCookie(ctx, {
            session,
            user: newUser
          });
          return ctx.json({
            token: session.token,
            user: {
              id: newUser.id,
              email: newUser.email,
              emailVerified: newUser.emailVerified,
              name: newUser.name,
              createdAt: newUser.createdAt,
              updatedAt: newUser.updatedAt
            }
          });
        }
      )
    },
    hooks: {
      after: [
        {
          matcher(context) {
            const setCookie = context.responseHeader.get("set-cookie");
            setCookie?.includes(
              context.context.authCookies.sessionToken.name
            );
            return context.path.startsWith("/sign-in") || context.path.startsWith("/sign-up") || context.path.startsWith("/callback") || context.path.startsWith("/oauth2/callback") || context.path.startsWith("/magic-link/verify") || context.path.startsWith("/email-otp/verify-email");
          },
          handler: chunkFTP5NPIZ_cjs.createAuthMiddleware(async (ctx) => {
            const headers = ctx.context.returned instanceof chunkFTP5NPIZ_cjs.APIError ? ctx.context.returned.headers : ctx.responseHeader;
            const setCookie = headers.get("set-cookie");
            const sessionTokenName = ctx.context.authCookies.sessionToken.name;
            const sessionCookie = chunkJOH6QBAP_cjs.parseSetCookieHeader(setCookie || "").get(sessionTokenName)?.value.split(".")[0];
            if (!sessionCookie) {
              return;
            }
            const session = await chunkFTP5NPIZ_cjs.getSessionFromCtx(
              ctx,
              {
                disableRefresh: true
              }
            );
            if (!session || !session.user.isAnonymous) {
              return;
            }
            if (ctx.path === "/sign-in/anonymous") {
              throw new chunkFTP5NPIZ_cjs.APIError("BAD_REQUEST", {
                message: ERROR_CODES.ANONYMOUS_USERS_CANNOT_SIGN_IN_AGAIN_ANONYMOUSLY
              });
            }
            const newSession = ctx.context.newSession;
            if (!newSession) {
              return;
            }
            if (options?.onLinkAccount) {
              await options?.onLinkAccount?.({
                anonymousUser: session,
                newUser: newSession
              });
            }
            if (!options?.disableDeleteAnonymousUser) {
              await ctx.context.internalAdapter.deleteUser(session.user.id);
            }
          })
        }
      ]
    },
    schema: chunkUQC2FUXC_cjs.mergeSchema(schema, options?.schema),
    $ERROR_CODES: ERROR_CODES
  };
};

exports.anonymous = anonymous;
