// better-auth/dist/cookies.cjs
'use strict';

var chunkJOH6QBAP_cjs = require('./chunk-JOH6QBAP.cjs');
require('./chunk-OA7Z7IZT.cjs');
require('./chunk-2HPSCSV7.cjs');
require('./chunk-VXYIYABQ.cjs');
require('./chunk-PEZRSDZS.cjs');



Object.defineProperty(exports, "createCookieGetter", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.createCookieGetter; }
});
Object.defineProperty(exports, "deleteSessionCookie", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.deleteSessionCookie; }
});
Object.defineProperty(exports, "getCookieCache", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.getCookieCache; }
});
Object.defineProperty(exports, "getCookies", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.getCookies; }
});
Object.defineProperty(exports, "getSessionCookie", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.getSessionCookie; }
});
Object.defineProperty(exports, "parseCookies", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.parseCookies; }
});
Object.defineProperty(exports, "parseSetCookieHeader", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.parseSetCookieHeader; }
});
Object.defineProperty(exports, "setCookieCache", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.setCookieCache; }
});
Object.defineProperty(exports, "setSessionCookie", {
  enumerable: true,
  get: function () { return chunkJOH6QBAP_cjs.setSessionCookie; }
});
