// better-auth/dist/db.cjs
'use strict';

var chunkBCGZXLBF_cjs = require('./chunk-BCGZXLBF.cjs');
var chunkUQC2FUXC_cjs = require('./chunk-UQC2FUXC.cjs');
require('./chunk-H74YRRNV.cjs');
require('./chunk-5E75URIA.cjs');
require('./chunk-NIMYOIVU.cjs');
require('./chunk-CCKQSGIR.cjs');
require('./chunk-G2LZ73E2.cjs');
require('./chunk-OA7Z7IZT.cjs');
require('./chunk-2HPSCSV7.cjs');
require('./chunk-VXYIYABQ.cjs');
require('./chunk-PEZRSDZS.cjs');



Object.defineProperty(exports, "convertFromDB", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.convertFromDB; }
});
Object.defineProperty(exports, "convertToDB", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.convertToDB; }
});
Object.defineProperty(exports, "createFieldAttribute", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.createFieldAttribute; }
});
Object.defineProperty(exports, "createInternalAdapter", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.createInternalAdapter; }
});
Object.defineProperty(exports, "getAdapter", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.getAdapter; }
});
Object.defineProperty(exports, "getAuthTables", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.getAuthTables; }
});
Object.defineProperty(exports, "getMigrations", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.getMigrations; }
});
Object.defineProperty(exports, "getSchema", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.getSchema; }
});
Object.defineProperty(exports, "getWithHooks", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.getWithHooks; }
});
Object.defineProperty(exports, "matchType", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.matchType; }
});
Object.defineProperty(exports, "toZodSchema", {
  enumerable: true,
  get: function () { return chunkBCGZXLBF_cjs.toZodSchema; }
});
Object.defineProperty(exports, "accountSchema", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.accountSchema; }
});
Object.defineProperty(exports, "getAllFields", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.getAllFields; }
});
Object.defineProperty(exports, "mergeSchema", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.mergeSchema; }
});
Object.defineProperty(exports, "parseAccountInput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseAccountInput; }
});
Object.defineProperty(exports, "parseAccountOutput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseAccountOutput; }
});
Object.defineProperty(exports, "parseAdditionalUserInput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseAdditionalUserInput; }
});
Object.defineProperty(exports, "parseInputData", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseInputData; }
});
Object.defineProperty(exports, "parseOutputData", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseOutputData; }
});
Object.defineProperty(exports, "parseSessionInput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseSessionInput; }
});
Object.defineProperty(exports, "parseSessionOutput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseSessionOutput; }
});
Object.defineProperty(exports, "parseUserInput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseUserInput; }
});
Object.defineProperty(exports, "parseUserOutput", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.parseUserOutput; }
});
Object.defineProperty(exports, "sessionSchema", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.sessionSchema; }
});
Object.defineProperty(exports, "userSchema", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.userSchema; }
});
Object.defineProperty(exports, "verificationSchema", {
  enumerable: true,
  get: function () { return chunkUQC2FUXC_cjs.verificationSchema; }
});
