// better-auth/dist/oauth2.cjs
'use strict';

var chunkV6RGXSG4_cjs = require('./chunk-V6RGXSG4.cjs');
var chunk5E75URIA_cjs = require('./chunk-5E75URIA.cjs');
require('./chunk-G2LZ73E2.cjs');
require('./chunk-2HPSCSV7.cjs');



Object.defineProperty(exports, "createAuthorizationURL", {
  enumerable: true,
  get: function () { return chunkV6RGXSG4_cjs.createAuthorizationURL; }
});
Object.defineProperty(exports, "generateCodeChallenge", {
  enumerable: true,
  get: function () { return chunkV6RGXSG4_cjs.generateCodeChallenge; }
});
Object.defineProperty(exports, "getOAuth2Tokens", {
  enumerable: true,
  get: function () { return chunkV6RGXSG4_cjs.getOAuth2Tokens; }
});
Object.defineProperty(exports, "validateAuthorizationCode", {
  enumerable: true,
  get: function () { return chunkV6RGXSG4_cjs.validateAuthorizationCode; }
});
Object.defineProperty(exports, "validateToken", {
  enumerable: true,
  get: function () { return chunkV6RGXSG4_cjs.validateToken; }
});
Object.defineProperty(exports, "generateState", {
  enumerable: true,
  get: function () { return chunk5E75URIA_cjs.generateState; }
});
Object.defineProperty(exports, "parseState", {
  enumerable: true,
  get: function () { return chunk5E75URIA_cjs.parseState; }
});
