// better-auth/dist/social.cjs
'use strict';

var chunkNU4SXYJ5_cjs = require('./chunk-NU4SXYJ5.cjs');
require('./chunk-V6RGXSG4.cjs');
require('./chunk-H74YRRNV.cjs');
require('./chunk-5E75URIA.cjs');
require('./chunk-CCKQSGIR.cjs');
require('./chunk-G2LZ73E2.cjs');
require('./chunk-2HPSCSV7.cjs');
require('./chunk-PEZRSDZS.cjs');



Object.defineProperty(exports, "SocialProviderListEnum", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.SocialProviderListEnum; }
});
Object.defineProperty(exports, "apple", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.apple; }
});
Object.defineProperty(exports, "discord", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.discord; }
});
Object.defineProperty(exports, "dropbox", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.dropbox; }
});
Object.defineProperty(exports, "facebook", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.facebook; }
});
Object.defineProperty(exports, "getApplePublicKey", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.getApplePublicKey; }
});
Object.defineProperty(exports, "github", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.github; }
});
Object.defineProperty(exports, "gitlab", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.gitlab; }
});
Object.defineProperty(exports, "google", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.google; }
});
Object.defineProperty(exports, "linkedin", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.linkedin; }
});
Object.defineProperty(exports, "microsoft", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.microsoft; }
});
Object.defineProperty(exports, "reddit", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.reddit; }
});
Object.defineProperty(exports, "socialProviderList", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.socialProviderList; }
});
Object.defineProperty(exports, "socialProviders", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.socialProviders; }
});
Object.defineProperty(exports, "spotify", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.spotify; }
});
Object.defineProperty(exports, "twitch", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.twitch; }
});
Object.defineProperty(exports, "twitter", {
  enumerable: true,
  get: function () { return chunkNU4SXYJ5_cjs.twitter; }
});
