// better-auth/dist/plugins/access.cjs
'use strict';

var chunk2X5G64P2_cjs = require('../chunk-2X5G64P2.cjs');



Object.defineProperty(exports, "adminAc", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.adminAc; }
});
Object.defineProperty(exports, "createAccessControl", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.createAccessControl; }
});
Object.defineProperty(exports, "defaultAc", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.defaultAc; }
});
Object.defineProperty(exports, "defaultRoles", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.defaultRoles; }
});
Object.defineProperty(exports, "defaultStatements", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.defaultStatements; }
});
Object.defineProperty(exports, "memberAc", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.memberAc; }
});
Object.defineProperty(exports, "ownerAc", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.ownerAc; }
});
Object.defineProperty(exports, "role", {
  enumerable: true,
  get: function () { return chunk2X5G64P2_cjs.role; }
});
